src/Servers/Kestrel/Core/src/Internal/Http2/FlowControl/StreamOutputFlowControl.cs

```diff
@@ -24,6 +24,12 @@ public StreamOutputFlowControl(OutputFlowControl connectionLevelFlowControl, uin
 
         public bool IsAborted => _connectionLevelFlowControl.IsAborted || _streamLevelFlowControl.IsAborted;
 
+        public void Reset()
+        {
+            _streamLevelFlowControl.Reset();
+            Debug.Assert(_currentConnectionLevelAwaitable == null, "Should have been completed and nulled by the previous stream.");
```

Comment:
This assert is invalid. AdvanceUpToAndWait is the only method that ever nulls _currentConnectionLevelAwaitable out, and there's no guarantee that the last call to AdvanceUpToAndWait doesn't just recreate it.

I would still at least verify _currentConnectionLevelAwaitable is completed (if it exists) then null it out. It's better to just throw in this case rather than do a Debug.Assert, because doing something like `await Task.WhenAny(Response.Body.WriteAsync(), Task.Delay(1000))` and then throwing when the Task.Delay completes first could lead to the stream being pooled despite there still being an ongoing call to WriteAsync.

This can happen because in the case an app throws an exception after it has started writing to the response body, [HttpProtocol skips the call](https://github.com/dotnet/aspnetcore/blob/ce0579255104f6ea1902b62c1e21195a7b93f828/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L1030-L1038) to IHttpOutputProducer.WriteStreamSuffixAsync which would normally be what delays the call to Http2Stream.OnRequestProcessingEnded until after Http2OutputProducer.ProcessDataWrites completes.

I'm starting to think we should not pool streams where the last app func exited with an Exception. In general, output flow control is an area that I'm very concerned about pooling, because tasks returned from Response.Body.WriteAsync() are often not observed.

Reply:
Call?